### ISSUE TITLE
Topic proposal

### ISSUE THREAD

**Issue by Garrafao** [CastIehard/LLM-Grammaticalization-Prediction#1]
**Introduction**
Grammaticalization is a process of language change where expressions become more more functional over time. This has effects on multiple linguistic levels, such as semantics, morphology or syntax. [1] predict the grammaticalization level of expressions using the German preposition dataset of [2]. [1] observed frequency to be the best predictor of grammaticalization, of three cooccurrence-frequency-based measures tested. Measures were calculated on a German web corpus. We try to outperform the frequency measure with LLM-based measures. We hypothesize that LLMs encoding information from multiple linguistic levels can help to predict grammaticalization.

**Related work**

[1] https://arxiv.org/abs/1804.06719
[2] https://esv-elibrary.de/journal/article/10.37307/j.2198-2430.2003.02.13
[3] https://www.degruyterbrill.com/document/doi/10.1515/9783110753073/html?lang=de
[4] https://arxiv.org/abs/2402.03927

This is not exhaustive, please do a search yourself, papers above also contain important references. Please search for references using LLMs for research on linguistic phenomena.

**Task**
The task is given a set of usages of an expression $e$, predict its grammaticalization level. An alternative setup gives only the target word as input.

**Data**
[2] induced grammaticalization levels roughly 20 years ago. So, it would be good to use a corpus from around that time. But these levels are (presumably) quite stable as grammaticalization happens slowly. I propose to use one of these corpora:
- [SdeWaC](https://www.ims.uni-stuttgart.de/forschung/ressourcen/korpora/sdewac/)
- appropriate time slice from [Leipzig corpus](https://corpora.uni-leipzig.de/de?corpusId=deu_news_2024), should be large enough, tbd

From the corpus you will need to extract usages (sentences) for expressions from [1]'s testset. This contains multi-words. Don't under-estimate this task. It requires some thinking.

Test set from [1], split into dev and test randomly.

**Models** 
Any LLM with
- systematic, documented prompt engineering 
- chain-of-thought prompting
- few/zero-shot
- [DSPy](https://dspy.ai/). This library creates optimized prompts from labeled data, kind of backwards engineering. The documentation is cryptic, but it works and can bring huge performance boosts. [This course](https://youtu.be/f5ym8e1pwHA?si=L7IK8vRZ2y8as4YN) is a good starting point. You will probably have to use this with a local instance of LLMs.
- Integrating the frequency information into the prompt.
- More?

Baselines:
- frequency calculated from corpus
- most frequent class baseline
- random baseline with different class distributions
- a BERT-based classifier (encoder-only model). Usually works better than LLMs. For this we would need to create a train split from [1].

**Evaluation**
Metrics defined in [1].

**Experiments**
*Important*: At the beginning, I suggest *not* to aim for breadth in models. Choose one "small" LLM, e.g. FLAN-T5 and implement the whole evaluation pipeline with it. Report results on github to sanity check. 

Validation experiment: Reproduce result from [1] with frequency calculated from our corpus. Report results on github.

Main experiments: tbd

Additional experiments to exclude data leakage: Many models were trained after [1]'s dataset was published. Hence, models could have seen gold data [4]. TODO: Think of experiments to test for this possibility.

**Conclusion**
Did it work?

**Comment by CastIehard** [comment]
Hi @Garrafao ,

I’m a bit confused about the corpus selection for our project. In your paper [1], you mention using the SdeWaC corpus to calculate word entropy. However, the GitHub repository linked in [1]’s footnote ([link here](https://github.com/Garrafao/MetaphoricChange)) suggests using the DTA corpus (DTA-Kernkorpus and Ergänzungstexte, TCF version from May 11, 2016) to reproduce results.

Since the DTA corpus is publicly available and already POS-tagged, it seems ideal for our purposes. Could you please clarify if using the DTA corpus would be appropriate for our analysis, or if we should stick with the SdeWaC corpus? I noticed the GitHub repository references a different paper focusing on metaphorical change rather than grammaticalization—is this corpus possibly unrelated to our current task?

I requested the SdeWaC Dataset already (put you CC in Mail), because I want to save time and iteration, and because my gut tells me there is a reason you didn't suggest us using DTA Dataset.

Thanks for your help!

Best regards,
Luca

**Comment by Garrafao** [comment]
> I’m a bit confused about the corpus selection for our project. In your paper [1], you mention using the SdeWaC corpus to calculate word entropy. However, the GitHub repository linked in [1]’s footnote ([link here](https://github.com/Garrafao/MetaphoricChange)) suggests using the DTA corpus (DTA-Kernkorpus and Ergänzungstexte, TCF version from May 11, 2016) to reproduce results.

In that paper, I just lazily linked the repository from another paper (where we used DTA) because for the grammaticalization paper I used exactly the same scripts for entropy, frequency, etc., but on another corpus.

> Since the DTA corpus is publicly available and already POS-tagged, it seems ideal for our purposes. Could you please clarify if using the DTA corpus would be appropriate for our analysis, or if we should stick with the SdeWaC corpus? I noticed the GitHub repository references a different paper focusing on metaphorical change rather than grammaticalization—is this corpus possibly unrelated to our current task?

The DTA corpus is a historical-diachronic corpus. It only contains data up to mid 20th-century. Most of the data is from earlier centuries. For this corpus, the grammaticalization levels we have are probably slightly inaccurate. I still think SdeWaC would be ideal. But, if we cannot get it, maybe DTA is an option.

> I requested the SdeWaC Dataset already (put you CC in Mail), because I want to save time and iteration, and because my gut tells me there is a reason you didn't suggest us using DTA Dataset.

Thanks for putting me in CC. But, you mentioned a "STE-WRC" data set, which is unknown to me. Maybe that was intended, just checking.

**Comment by CastIehard** [comment]
Thank you for your incredibly quick reply!

I understand it now. I was so confused before, but it all makes sense now!

... how embarrassing! I meant to say the **SdeWaC** dataset. I ran my email through DeepL Write and didn't notice the change. I will send them an additional email specifying the dataset.

I will start working on the DTA dataset to get a feel for working with tagged data. If we can get the SdeWaC dataset within the next week, I think we can switch by changing a few lines of code.

Have a nice weekend, and thanks for the help!

**Comment by Garrafao** [comment]
I reopened this because the tasks defined in the topic proposal are not implemented yet. Please only close this once we have a concrete working plan in another issue for each of you that implements the above-described topic proposal.

**Comment by CastIehard** [comment]
# Meeting 30.07.2025 11 a.m

Participants: @Garrafao @CastIehard @prasanna9481 @divyansh-kaushik1

## Summary & To-Do List

### Evaluation & Data @CastIehard 

* Create a list of keywords in both dev and test sets
* Compute dev results
* Split keyword metrics into dev and test
* Create a general evaluation script to automate metric calculation
* Final goal: one big results table with all evaluations
* Split dev set into dev1 and dev2 (for prompting)
* Document how the metric calculations work

### Baselines & Prompts @prasanna9481 

* Baseline: use only the word (no sentence) and ask for grammaticalization level
* Do this with and without definitions
* Prompt variations: a, b, c, d… increasing in complexity, fine-grained
* Add a config file (JSON) for prompt settings:

  * prompt: string
  * few\_shot: boolean
  * sentence\_level: boolean
  * ... (so all prompts are plug and play)

* Don’t start DSPy yet – only after simple setup works
* Later use DSPy with e.g. labeled few-shot examples
* Describe how your prompts were created (e.g., Expected format: label: — where does this come from?)
* Also document DSPy options and what it optimizes (e.g., “give 10 examples and refine prompt”)

* Start with expressions only, sentences are just a bonus
* @prasanna9481’s script: takes keywords + grammar score → outputs predictions in new column
* @CastIehard’s script: takes that file → computes all evaluation metrics
* Add references for where definitions came from (e.g., entropy, second-order entropy) — cite the source paper

### Data Leakage @divyansh-kaushik1 

* Investigate: Was the test set part of LLM training data?
* Completing a word like a preposition only shows grammatical knowledge, not necessarily leakage
* Give e.g. 10 examples from the test set and see if the LLM completes the rest of the test set based on these
* Check what set of prepositions we expect in test. 
* Test with completion mode also (inserting e.g. 50 percent of Testset and see if it can complete it)
* Look at e.g. Bert (any llm were publishers puplished the training data) publishing and check if data was part of training
* Find data that was in fact part of training and check the methods with it


Submit final version of Report to @Garrafao on Wed 13.08.2025

**Comment by Garrafao** [comment]
Remotely related work:

https://escholarship.org/uc/item/7hr2d51f

**Comment by CastIehard** [comment]
Hi together,

_I did multiple pull requests and a lot of changes — here is a single message that summarizes everything to avoid confusion._

As discussed, I’ve implemented the following action items:

> Create a list of keywords in both dev and test sets
> Split dev set into dev1 and dev2 (for prompting)

---

## Improved Dataset Split

**I chose a better approach:**
Before splitting, I set aside **8 keywords** (2 randomly chosen for each grammaticalization level) for potential few-shot or one-shot prompting (**dev1 data**). The remaining 198 keywords were then split 50/50 into:

* **dev2 data** – for testing during development
* **test data** – for final evaluation

This way, we ensure clear separation between development and final testing and avoid data leakage.

#### DevData1 (for prompting)

Contains the 8 held-out keywords (2 per grammaticalization level) used for few-shot or one-shot prompt engineering.

#### DevData2 (for development-time testing)

Used during prompt iteration and validation steps. Metrics have been calculated specifically for this subset.

#### TestData (final evaluation only)

Strictly held out for final evaluation. No touching until everything is finalized.

Here's a screenshot of the new folder structure:

<img width="323" height="155" alt="Image" src="https://github.com/user-attachments/assets/44c98f88-52fa-4d9b-bb65-b97522ccb065" />

The folder 'general data' just contains the ground truth and statistics about the corpus and the dataset.

---

> Compute dev results

Evaluation metrics have been computed for all datasets independently. The 'dev2', 'test' and 'full data' folders in the screenshot above have all a file called '..._metrics.csv' that contains the same metrics as before, but now specific for the dataset as requested.

> Split keyword metrics into dev and test

Each folder now also contains a file with the keywords file. Here is a explanation:

<img width="319" height="134" alt="Image" src="https://github.com/user-attachments/assets/cd730c4d-a5be-433e-ab95-37195939c33a" />

The folder contains the dataset in the .jsonl file, the keywords with their metrics in the same named .csv file, the evaluation table with the metrics from before and an png with the grammaticalization level distribution:

<img width="500" height="400" alt="Image" src="https://github.com/user-attachments/assets/e249f8a6-0901-41cd-9b3e-478ab506b29c" />

## Most important part for @prasanna9481 and @divyansh-kaushik1 : Evaluation Pipeline
> Create a general evaluation script to automate metric calculation

A general-purpose evaluation script is now available. It can automatically compute key metrics (Spearman’s ρ, Average Precision) for any file added to `evaluation/input_csv/`.

> Final goal: one big results table with all evaluations

Implemented. All evaluated runs are compiled into one CSV and plotted in the same figure. Just drop your CSV into `evaluation/input_csv/` and re-run. @prasanna9481 and @divyansh-kaushik1 can simply add a .csv file with to the folder for each method or attempt and done. Here is an explanation on how it works (from 'evaluations/readme.md')

<img width="543" height="879" alt="Image" src="https://github.com/user-attachments/assets/bbcd0fbe-cfce-49ae-82dd-d6d99236df26" />

Let me know if anything is unclear. Happy to help anyone integrate their results — the new structure should make it easier for everyone to contribute!
